bluephos/bluephos_pipeline.py

```diff
@@ -1,77 +1,195 @@
 
 import pandas as pd
+import os
+import json
 from dplutils import cli
 from dplutils.pipeline.ray import RayStreamGraphExecutor
 
 from bluephos.tasks.generateligandtable import GenerateLigandTableTask
 from bluephos.tasks.nn import NNTask
 from bluephos.tasks.optimizegeometries import OptimizeGeometriesTask
 from bluephos.tasks.smiles2sdf import Smiles2SDFTask
-from bluephos.tasks.dft_orca import DFTTask
+from bluephos.tasks.dft import DFTTask
+
+
+def initialize_dataframe():
+    """Initialize a DataFrame with the required columns."""
+    columns = [
+        "ligand_identifier",
+        "ligand_SMILES",
+        "halide_identifier",
+        "halide_SMILES",
+        "acid_identifier",
+        "acid_SMILES",
+        "structure",
+        "z",
+        "xyz",
+        "ste",
+        "energy diff",
+    ]
+    return pd.DataFrame(columns=columns)
 
 
 def ligand_pair_generator(halides_file, acids_file):
+    """
+    Generate ligand pairs from halides and acids files.
+
+    Args:
+        halides_file (str): Path to the CSV file containing halides data.
+        acids_file (str): Path to the CSV file containing acids data.
+
+    Yields:
+        DataFrame: A DataFrame with a single row representing a ligand pair.
+    """
     halides_df = pd.read_csv(halides_file)
     acids_df = pd.read_csv(acids_file)
     for _, halide in halides_df.iterrows():
         for _, acid in acids_df.iterrows():
-            ligand_pair = {
-                "halide_identifier": halide["halide_identifier"],
-                "halide_SMILES": halide["halide_SMILES"],
-                "acid_identifier": acid["acid_identifier"],
-                "acid_SMILES": acid["acid_SMILES"],
-            }
-            yield (pd.DataFrame([ligand_pair]))
+            ligand_pair_df = initialize_dataframe()
+
+            ligand_pair_df.at[0, "halide_identifier"] = halide["halide_identifier"]
+            ligand_pair_df.at[0, "halide_SMILES"] = halide["halide_SMILES"]
+            ligand_pair_df.at[0, "acid_identifier"] = acid["acid_identifier"]
+            ligand_pair_df.at[0, "acid_SMILES"] = acid["acid_SMILES"]
+
+            yield ligand_pair_df
+
+
+def rerun_candidate_generator(input_dir, t_nn, t_ste, t_ed):
+    """
+    Generates candidate DataFrames from parquet files in the input directory.
+
+    Core Algorithm:
+    - If the absolute value of 'z' is less than t_nn,
+    - and 'ste' is None or its absolute value is less than t_ste,
+    - and 'energy diff' is None,
+    This row is then added to a new DataFrame and yielded for re-run.
+
+    Additional Context:
+    1. All valid ligand pairs should already have run through the NN process and have a 'z' score.
+    2. If a row's 'ste' is None, then it's 'energy diff' should also be None.
+
+    Args:
+        input_dir (str): Directory containing input parquet files.
+        t_nn (float): Threshold for 'z' score.
+        t_ste (float): Threshold for 'ste'.
+        t_ed (float): Threshold for 'energy diff'.
```

Comment:
Yes, t_ed is not currently used since it's part of the final step. I keep the score in the results to avoid recalculating it in the future. we may need to ask Alexander about how to use it.

Reply:
yeah, that is the core deliverable. Clearly we need the energy diff in output, but the threshold seems outside the pipeline. For example, after a run they can consume the results and filter any given way to decide what to manufacture